File: 3600-3699/3612. Process String with Special Operations I/Solution.cpp
// Approach-1 (String Simulation)
// T.C : O(n²)
// S.C : O(n)

class Solution
{
public:
  string processStr(string s)
  {

    string result;

    for (char ch : s)
    {

      if (ch == '*')
      {

        if (!result.empty())
        {

          result.pop_back();
        }
      }
      else if (ch == '#')
      {

        result += result;
      }
      else if (ch == '%')
      {

        reverse(result.begin(), result.end());
      }
      else
      {

        result.push_back(ch);
      }
    }

    return result;
  }
};

/*
//Approach-2 (Vector Simulation)
//T.C : O(n²)
//S.C : O(n)

class Solution {
public:

    string processStr(string input) {

        vector<char> characters;

        for (char current : input) {

            if (current >= 'a' && current <= 'z') {

                characters.push_back(current);
            }
            else if (current == '*') {

                if (!characters.empty()) {

                    characters.pop_back();
                }
            }
            else if (current == '#') {

                int size = characters.size();

                for (int i = 0; i < size; i++) {

                    characters.push_back(characters[i]);
                }
            }
            else {

                reverse(
                    characters.begin(),
                    characters.end()
                );
            }
        }

        string answer;

        for (char ch : characters) {

            answer += ch;
        }

        return answer;
    }
};
*/

/*
//Approach-3 (StringBuilder Equivalent)
//T.C : O(n²)
//S.C : O(n)

class Solution {
public:

    string processStr(string input) {

        string output;

        for (char current : input) {

            if (islower(current)) {

                output.push_back(current);
            }
            else if (current == '*') {

                if (!output.empty()) {

                    output.pop_back();
                }
            }
            else if (current == '#') {

                output += output;
            }
            else {

                reverse(
                    output.begin(),
                    output.end()
                );
            }
        }

        return output;
    }
};
*/

File: 3600-3699/3612. Process String with Special Operations I/Solution.java
//Approach-1 (StringBuilder Simulation)
//T.C : O(n²) Worst Case
//S.C : O(n)

class Solution {

  public String processStr(String s) {

    StringBuilder sb = new StringBuilder();

    for (char ch : s.toCharArray()) {

      if (ch == '*') {

        if (sb.length() > 0) {
          sb.deleteCharAt(sb.length() - 1);
        }
      } else if (ch == '#') {

        StringBuilder duplicate = new StringBuilder(sb);

        sb.append(duplicate);
      } else if (ch == '%') {

        sb.reverse();
      } else {

        sb.append(ch);
      }
    }

    return sb.toString();
  }
}

/*
 * //Approach-2 (ArrayList Simulation)
 * //T.C : O(n²)
 * //S.C : O(n)
 * 
 * class Solution {
 * 
 * public String processStr(String input) {
 * 
 * List<Character> chars = new ArrayList<>();
 * 
 * for (char current : input.toCharArray()) {
 * 
 * if (current >= 'a' && current <= 'z') {
 * 
 * chars.add(current);
 * }
 * else if (current == '*') {
 * 
 * if (!chars.isEmpty()) {
 * 
 * chars.remove(chars.size() - 1);
 * }
 * }
 * else if (current == '#') {
 * 
 * int size = chars.size();
 * 
 * for (int i = 0; i < size; i++) {
 * 
 * chars.add(chars.get(i));
 * }
 * }
 * else {
 * 
 * Collections.reverse(chars);
 * }
 * }
 * 
 * StringBuilder answer = new StringBuilder();
 * 
 * for (char ch : chars) {
 * 
 * answer.append(ch);
 * }
 * 
 * return answer.toString();
 * }
 * }
 */

/*
 * //Approach-3 (Optimized StringBuilder)
 * //T.C : O(n²)
 * //S.C : O(n)
 * 
 * class Solution {
 * 
 * public String processStr(String data) {
 * 
 * StringBuilder output = new StringBuilder();
 * 
 * for (char current : data.toCharArray()) {
 * 
 * if (Character.isLowerCase(current)) {
 * 
 * output.append(current);
 * }
 * else if (current == '*') {
 * 
 * if (output.length() > 0) {
 * 
 * output.deleteCharAt(output.length() - 1);
 * }
 * }
 * else if (current == '#') {
 * 
 * output.append(output.toString());
 * }
 * else {
 * 
 * output.reverse();
 * }
 * }
 * 
 * return output.toString();
 * }
 * }
 */